front/app/containers/EventsPage/EventsViewer/index.tsx

```diff
@@ -66,57 +64,47 @@ const EVENTS_PER_PAGE = 10;
 const EventsViewer = memo<Props>(
   ({ title, fallbackMessage, eventsTime, className }) => {
     const [projectIds, setProjectIds] = useState<string[]>([]);
+    const [currentPageNumber, setCurrentPageNumber] = useState<number>(1);
 
-    const events = useEvents(
+    const { events, lastPageNumber } = useEvents({
       projectIds,
-      eventsTime === 'future',
-      eventsTime === 'past'
-    );
-
-    const eventsHaveLoaded = !isNilOrError(events);
-
-    const [currentPage, setCurrentPage] = useState<number>(1);
-    const [visibleEvents, setVisibleEvents] = useState<IEventData[]>([]);
-
-    useEffect(() => {
-      if (isNilOrError(events)) return;
-      setVisibleEvents(sliceEventsToPage(events, currentPage, EVENTS_PER_PAGE));
-    }, [events, currentPage]);
+      futureOnly: eventsTime === 'future',
+      pastOnly: eventsTime === 'past',
+      pageNumber: currentPageNumber,
+      pageSize: EVENTS_PER_PAGE,
+    });
 
     return (
       <div className={className}>
         <TopBar title={title} setProjectIds={setProjectIds} />
 
-        {!eventsHaveLoaded && <Spinner />}
+        {isNilOrError(events) && <Spinner />}
```

Comment:
Not exactly right. If there's an error, it will keep on spinning. You need to keep loading state here in the component, similar to what you did before, or maybe we can even have this in hooks, but that's out of scope.

Reply:
Let me know if you want me to have a stab at it. Might be faster than the back and forth in the comments :-) In any case, loadPage's function should first be changed (see below).